cmd/promtool/main.go

```diff
@@ -57,6 +64,17 @@ func main() {
 	updateRulesCmd := updateCmd.Command("rules", "Update rules from the 1.x to 2.x format.")
 	ruleFilesUp := updateRulesCmd.Arg("rule-files", "The rule files to update.").Required().ExistingFiles()
 
+	queryCmd := app.Command("query", "Run query against a Prometheus server.")
+	queryInstantCmd := queryCmd.Command("instant", "Run instant query.")
+	queryServer := queryInstantCmd.Arg("server", "Prometheus server to query.").Required().URL()
+	queryExpr := queryInstantCmd.Arg("expr", "PromQL query expression.").Required().String()
```

Comment:
`queryServer` -> `queryInstantServer`
`queryExpr` -> `queryInstantExpr`

Too bad identical subarguments are apparently not reusable between range and instant queries.

Reply:
I think  somehting like  this should  work.

```
queryCmd := app.Command("query", "Run query against a Prometheus server.")
queryServer := queryCmd.Arg("server", "Prometheus server to query.").Required().URL()
queryExpr := queryCmd.Arg("expr", "PromQL query expression.").Required().String()

queryInstantCmd := queryCmd.Command("instant", "Run instant query.")

queryRangeCmd := queryCmd.Command("range", "Run range query.")
queryRangeBegin := queryRangeCmd.Flag("start", "Query range start time (RFC3339 or Unix timestamp).").String()
queryRangeEnd := queryRangeCmd.Flag("end", "Query range end time (RFC3339 or Unix timestamp).").String()
```

Reply:
Yeah, we tried that earlier, but then kingpin complains with:

```
promtool: error: can't mix Arg()s with Command()s, try --help
```

Reply:
what about using it as a flag?
`queryServer := queryCmd.Flag...`

Reply:
That might work, but it'd be nice to be able to use it as a positional arg, since it's required anyway...

Reply:
Thinking out loud but maybe we could have a single command for both instant and range queries?

```
usage: main query [<flags>] <server> <expr>

Run query against a Prometheus server.

Flags:
  -h, --help         Show context-sensitive help (also try --help-long and --help-man).
      --version      Show application version.
      --start=START  Query range start time (RFC3339 or Unix timestamp).
      --end=END      Query range end time (RFC3339 or Unix timestamp).

Args:
  <server>  Prometheus server to query.
  <expr>    PromQL query expression.
```

When `start` and `end` are given, the command returns a range query otherwise it's an instance query.

Reply:
@simonpasquier But for an instant query you can also provide a timestamp, for which `start` or `end` wouldn't be appropriate naming, so then we'd have yet another flag and there'd be more combinatorial confusion going on. IMO it's better to keep the range/instant queries explicit and separate, as they also hit different API endpoints.